### ISSUE TITLE
FitLog Ex Equipment as Tag

### ISSUE THREAD

**Issue by wolfgang-ch** [wolfgang-ch/mytourbook#388]
@FJBDev 

This code is a little bit confusing me. 

https://github.com/wolfgang-ch/mytourbook/blob/ebddd32c9daed192cf96046992e747b8c234c2e3/bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java#L1253-L1322

I need to make the creation of tour tags concurrently and this code needs to be adjusted that tour tags are not created and saved during the import (which is also not OK until now, they should be saved in the db, when a tour is saved)

My Question:

When a tag is created for an equipment, can there be multiple tags with the same name but with different equipment id's in the tag notes?

When yes, this is not a good idea

**Comment by FJBDev** [comment]
> When a tag is created for an equipment, can there be multiple tags with the same name but with different equipment id's in the tag notes?
> 
> When yes, this is not a good idea

Yes, it's possible for multiple tags to have the same name.
I just did a test in SportTracks 3.1. I created 2 tags with the same name `Tag1 `and assigned it to an activity, here is what the exported Fitlog file looks like

```
 <EquipmentUsed>
    <EquipmentItem Id="53207278-4ba0-4238-aed6-0d3015299800" Name="Tag1" />
    <EquipmentItem Id="e944b696-1e5b-4d88-a0f0-e3598360e64d" Name="Tag1" />
   </EquipmentUsed>
```

**Comment by wolfgang-ch** [comment]
Since the beginning, the whole tagging system in MT is based only on a unique tag name for all tags, e.g. in the "Tagged Tours" view you couldn't identify a tag, when multiple tags would have the same name.

In the FitLog importer, I'll remove the dependency on the tag notes content when a tag is searched, if it is available or not

**Comment by FJBDev** [comment]
I'm not sure to understand the issue and the need for a solution since in MT, each tag has a [tagId ](https://github.com/wolfgang-ch/mytourbook/blob/ebddd32c9daed192cf96046992e747b8c234c2e3/bundles/net.tourbook/src/net/tourbook/data/TourTag.java#L58)so the example above from Sporttracks shouldn't be a problem no ?

In MT, I have tags with the same name and it works fine.

![image](https://user-images.githubusercontent.com/25235665/131543000-4cb0bd65-ab84-4158-b97c-052a10112439.png)

**Comment by wolfgang-ch** [comment]
I'll reimplement it but first the simpel version has to work

**Comment by wolfgang-ch** [comment]
I discovered, when `searchTourTagInDatabase()` is called, then `_equipments` is `null` because it is set in line 340, after the SAX parser has finished

https://github.com/wolfgang-ch/mytourbook/blob/b8cfcd6336e0396fde0cff0641da17bbb465e02e/bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java#L1300-L1309


https://github.com/wolfgang-ch/mytourbook/blob/b8cfcd6336e0396fde0cff0641da17bbb465e02e/bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java#L326-L343

**Comment by wolfgang-ch** [comment]
When importing the test file then I see, that this import creates two `Shoes 2` tags, with and without tag notes, this is what I expected with the above code because in `searchTourTagInDatabase` there is no equipment and it creates a tag without tag notes and the tour is attached to the tag without tag notes. So there seems to be some issues in the current main branch.

![image](https://user-images.githubusercontent.com/1283445/131796520-3db9d3df-ffa2-4d7e-8fc5-8b68abfb1ff5.png)

**Comment by wolfgang-ch** [comment]
I fixed this issue when there are many tags

**OLD**

![image](https://user-images.githubusercontent.com/1283445/131800318-4576828d-1ca6-4d99-8519-eecc03ae8fc3.png)


**NEW**

![image](https://user-images.githubusercontent.com/1283445/131800338-c250ecd1-9a31-4895-83eb-2d0f4cc1c835.png)

**Comment by FJBDev** [comment]
> 
> 
> I discovered, when `searchTourTagInDatabase()` is called, then `_equipments` is allways `null` because it is set in line 340, after the SAX parser has finished
> 
> https://github.com/wolfgang-ch/mytourbook/blob/b8cfcd6336e0396fde0cff0641da17bbb465e02e/bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java#L1300-L1309
> 
> https://github.com/wolfgang-ch/mytourbook/blob/b8cfcd6336e0396fde0cff0641da17bbb465e02e/bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java#L326-L343

But there was a reason why it's set to null. 

I added it to 21.6 as it prevented to re-import and duplicate the tags and tour types

![image](https://user-images.githubusercontent.com/25235665/131915114-13a6f72c-0569-4666-a45b-1be6045981ab.png)

**Comment by FJBDev** [comment]
> 
> 
> When importing the test file then I see, that this import creates two `Shoes 2` tags, with and without tag notes, this is what I expected with the above code because in `searchTourTagInDatabase` there is no equipment and it creates a tag without tag notes and the tour is attached to the tag without tag notes. So there seems to be some issues in the current main branch.
> 

When I import the test file, I only see 1 tag "Shoes 2".
I think I'm still confused and don't understand the issue here

![image](https://user-images.githubusercontent.com/25235665/131915631-2a4542bf-ee00-4b10-b2eb-384a6a20edc1.png)

**Comment by wolfgang-ch** [comment]
> When I import the [test file](https://github.com/wolfgang-ch/mytourbook/blob/main/bundles/net.tourbook.tests/test/importdata/sporttracks/fitlog/files/TimothyLake.fitlog) with a fresh 21.6.1, I only see 1 tag "Shoes 2".
> I think I'm still confused and don't understand the issue here

In your screenshot you can see, that the tag notes is not created, but it should

You have not saved the tour, then the 2nd tag is created, which is also confusing me, but, I'm currently rewriting this code to fix all these issues and the concurrency. Concurrency is already working without tag notes but tag notes make it much more complicated

**Tour Import**

![image](https://user-images.githubusercontent.com/1283445/131952816-15d1ee88-6199-4b44-a669-34d0b201905e.png)

**Tour Save**

![image](https://user-images.githubusercontent.com/1283445/131952866-0f2b7ad3-68f4-4e0f-a9d5-12df069e9fac.png)

**Easy Import**

![image](https://user-images.githubusercontent.com/1283445/131952918-7c527d04-dbd0-4f5e-8ee2-71e9fdbbde43.png)

**Comment by wolfgang-ch** [comment]
@FJBDev 
Running the import concurrently shows a new problem, which file wins when both FitLog and FitLogEx files are available


These are my test files with minor changes in number of track points, equipment and category (tour type)

![image](https://user-images.githubusercontent.com/1283445/132011256-8d36ae10-0d34-4628-9252-0071e05991bc.png)


This is the concurrent import result, yellow are the ex files. It is unpredictable which gets into the list, either FitLog or FitLogEx

![image](https://user-images.githubusercontent.com/1283445/132011318-246f2159-cf8d-4d03-ae77-6dafd86b233d.png)


I need a logic, for the selected import files:

Which files should be imported concurrently and which file is subimported and **MUST** not be imported concurrently ?

**Comment by FJBDev** [comment]
> In your screenshot you can see, that the tag notes is not created, but it should

No, it's normal the tag notes is not created because in the file, there are not tag notes

> 
> You have not saved the tour, then the 2nd tag is created, which is also confusing me, but, I'm currently rewriting this code to fix all these issues and the concurrency. Concurrency is already working without tag notes but tag notes make it much more complicated
> 

When, I save the tour, the 2nd tag is not created and it seems tat everything is working well

![image](https://user-images.githubusercontent.com/25235665/132066770-c680fb1f-4f5e-4126-94c4-dbe5d97ddeb2.png)

**Comment by FJBDev** [comment]
> @FJBDev
> Running the import concurrently shows a new problem, which file wins when both FitLog and FitLogEx files are available
> 
> These are my test files with minor changes in number of track points, equipment and category (tour type)
> 
> This is the concurrent import result, yellow are the ex files. It is unpredictable which gets into the list, either FitLog or FitLogEx
> 
> ![image](https://user-images.githubusercontent.com/1283445/132011318-246f2159-cf8d-4d03-ae77-6dafd86b233d.png)
> 
> I need a logic, for the selected import files:
> 
> Which files should be imported concurrently and which file is subimported and **MUST** not be imported concurrently ?

Well, technically, a user would not have 2 of the same file in both formats (Fitlog & FitlogEx). My experience is that it would be either or so that scenario should not happen. And If it were to happen, I would prioritize the FitLogEx format because it's superior to the Fitlog format in the sense that it has more data. [I've personally always recommended to use the FitlogEx format for these reasons](https://sourceforge.net/p/mytourbook/discussion/622811/thread/e9cfb35008/?limit=25#9eb4/5e69) and this is the file format I used to transfer all my activities from 1996 to 2018 in to MT.

**Comment by wolfgang-ch** [comment]
When there is only a FitLog or a FitLogEx file, then I don't see a need to subimport FitLogEx equipments within the FitLog parser.

When both files with the same name are available, then the FitLogEx file will be imported and the FitLog file is skipped.

This is my conclusion from your answer.

**Comment by wolfgang-ch** [comment]
>When there is only a FitLog or a FitLogEx file, then I don't see a need to subimport FitLogEx equipments within the FitLog parser.

This is already working as expected, it is not a fully subimport

**Comment by wolfgang-ch** [comment]
Closing this issue, it is implemented in MT 21.9
